tensorflow/contrib/rnn/python/ops/lstm_ops.py

```diff
@@ -572,9 +572,8 @@ def call(self, inputs, initial_state=None, dtype=None, sequence_length=None):
 
   def _gather_states(self, data, indices, batch_size):
     """Produce `out`, s.t. out(i, j) = data(indices(i), i, j)."""
-    mod_indices = indices * batch_size + math_ops.range(batch_size)
-    return array_ops.gather(
-        array_ops.reshape(data, [-1, self.num_units]), mod_indices)
+    return array_ops.gather_nd(
```

Comment:
What's the motivation for this?

Reply:
Because _gather_states find out(i, j) = data(indices(i), i, j) when indices is a [batch_size] tensor . The origin method  reshape the 'data' to a 2-dim tensor and calculate the indices in the reshaped array . Use gather_nd can locate the 'indices' in the 3-dim tensor 'data' so no need for reshape . And in my experiment , gather may lead to a warning when data is a sparse tensor but use gather_nd with no warning . And use gather_nd can speed up a little .

Reply:
OK, makes sense.